pushnotifications/src/main/java/com/pusher/pushnotifications/reporting/OpenNotificationActivity.kt

```diff
@@ -0,0 +1,86 @@
+package com.pusher.pushnotifications.reporting
+
+import android.app.Activity
+import android.content.Intent
+import android.os.Bundle
+import com.firebase.jobdispatcher.*
+import com.google.gson.Gson
+import com.google.gson.JsonSyntaxException
+import com.pusher.pushnotifications.logging.Logger
+import com.pusher.pushnotifications.reporting.api.ReportEvent
+import com.pusher.pushnotifications.reporting.api.ReportEventType
+
+/*
+ * This activity will be opened when a user taps a notification sent by the Pusher push notifications
+ * service. It reports back to Pusher that the notification has been opened and then
+ * opens the customer application activity that would have been opened had it not been
+ * intercepted.
+ */
+class OpenNotificationActivity: Activity() {
+    private val gson = Gson()
+    private val log = Logger.get(this::class)
+
+    private fun startIntent(clickAction: String? = null) {
+        val i: Intent
+        if (clickAction != null) {
+            i = Intent()
+            i.action = clickAction
+        } else {
+            i = packageManager.getLaunchIntentForPackage(packageName)
+        }
+
+        // We need to clear the activity stack so that this activity doesn't show up when customers
+        // are debugging.
+        i.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
+        try {
+            applicationContext.startActivity(i)
+        } catch (_: RuntimeException) {
+            log.e("Failed to start activity using clickAction $clickAction")
+            // The default Firebase behaviour in this situation is to kill the app, so we should
+            // do the same.
+            finishAffinity()
+        }
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        intent?.getStringExtra("pusher")?.let { pusherDataJson ->
+            try {
+                val pusherData = gson.fromJson(pusherDataJson, PusherMetadata::class.java)
```

Comment:
I'd maybe defer instantiation of Gson to here - it's the first place we can use it and if the intent extras aren't there it won't ever get to it.